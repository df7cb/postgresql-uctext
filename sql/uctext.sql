create extension uctext;

select null::uctext;
select ''::uctext;
select 'abc123'::uctext;
select 'ABC123'::uctext;

create table uc (
	u uctext,
	t text
);

select 'moo'::uctext = 'moo'::text, 'moo'::text = 'moo'::uctext;
select 'moo'::uctext = 'MOO'::text, 'MOO'::text = 'moo'::uctext;

insert into uc values ('df7cb', 'df7cb');
insert into uc values ('DF7CB', 'df7cb');

select u, t, u = t, t = u from uc;
select * from uc where u = 'df7cb';

insert into uc select 'abC'||i::text from generate_series(1, 10000) g(i);

create index on uc (u);
analyze uc;

select * from uc where u = 'abc123';
explain select * from uc where u = 'abc123';

select * from uc where u between 'abc123' and 'abc124';
