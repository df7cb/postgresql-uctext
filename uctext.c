/*
Copyright (C) 2021 Christoph Berg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/

#include <postgres.h>
#include <fmgr.h>
#include <utils/builtins.h>
#include <utils/formatting.h>

/* module initialization */

PG_MODULE_MAGIC;

/* functions */

PG_FUNCTION_INFO_V1 (uctext_in);

Datum
uctext_in (PG_FUNCTION_ARGS)
{
	char		*str = PG_GETARG_CSTRING(0);
	int			len = strlen(str);
	char		*out_string;
	text		*result;
	Oid         collid = PG_GET_COLLATION();

	out_string = str_toupper(str, len, collid);
	result = cstring_to_text(out_string);
	pfree(out_string);

	PG_RETURN_TEXT_P(result);
}

PG_FUNCTION_INFO_V1(uctext_out);

Datum
uctext_out(PG_FUNCTION_ARGS)
{
	Datum       txt = PG_GETARG_DATUM(0);

	PG_RETURN_CSTRING(TextDatumGetCString(txt));
}
