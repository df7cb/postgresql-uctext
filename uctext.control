default_version = '1'
comment = 'UPPER CASE text'
relocatable = true
trusted = true
