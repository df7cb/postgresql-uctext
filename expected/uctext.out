create extension uctext;
select null::uctext;
 uctext 
--------
 
(1 row)

select ''::uctext;
 uctext 
--------
 
(1 row)

select 'abc123'::uctext;
 uctext 
--------
 ABC123
(1 row)

select 'ABC123'::uctext;
 uctext 
--------
 ABC123
(1 row)

create table uc (
	u uctext,
	t text
);
select 'moo'::uctext = 'moo'::text, 'moo'::text = 'moo'::uctext;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

select 'moo'::uctext = 'MOO'::text, 'MOO'::text = 'moo'::uctext;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

insert into uc values ('df7cb', 'df7cb');
insert into uc values ('DF7CB', 'df7cb');
select u, t, u = t, t = u from uc;
   u   |   t   | ?column? | ?column? 
-------+-------+----------+----------
 DF7CB | df7cb | t        | t
 DF7CB | df7cb | t        | t
(2 rows)

select * from uc where u = 'df7cb';
   u   |   t   
-------+-------
 DF7CB | df7cb
 DF7CB | df7cb
(2 rows)

insert into uc select 'abC'||i::text from generate_series(1, 10000) g(i);
create index on uc (u);
analyze uc;
select * from uc where u = 'abc123';
   u    | t 
--------+---
 ABC123 | 
(1 row)

explain select * from uc where u = 'abc123';
                             QUERY PLAN                             
--------------------------------------------------------------------
 Index Scan using uc_u_idx on uc  (cost=0.29..8.30 rows=1 width=13)
   Index Cond: ((u)::text = 'ABC123'::text)
(2 rows)

select * from uc where u between 'abc123' and 'abc124';
    u    | t 
---------+---
 ABC123  | 
 ABC1230 | 
 ABC1231 | 
 ABC1232 | 
 ABC1233 | 
 ABC1234 | 
 ABC1235 | 
 ABC1236 | 
 ABC1237 | 
 ABC1238 | 
 ABC1239 | 
 ABC124  | 
(12 rows)

